ManageIQ/manageiq
# Move API OpenID-Connect support to Apache configuration

Fixes ManageIQ/manageiq#19866

This PR, combined with others in other github repos, will move the support for the ManageIQ API out of code and into the Apache configuration.

**Dependent PRs**
https://github.com/ManageIQ/manageiq-appliance/pull/282
https://github.com/ManageIQ/manageiq-api/pull/828
https://github.com/ManageIQ/manageiq-appliance_console/pull/117

**To test:**

Follow the ManageIQ OpenID-Connect documentation, found [here](https://www.manageiq.org/docs/reference/latest/auth/openid_connect), to configure an ManageIQ appliance for OpenID-Connect authentication and
confirm the API can be user with:

1. Basic auth using admin/<password>
2. A JWT acquired from the OpenID-Connect Identity Provider then pass the JWT in the header to make a request to the ManageIQ API
3. An API Auth token

Examples of each of the above 3:

1. Basic auth using admin/<password>

`curl -L -vvv -k --user admin:smartvm -X GET -H "Accept: application/json" https://${OID_CLIENT_HOST}/api/users `

2. A JWT acquired from the OpenID-Connect Identity Provider then pass the JWT in the header to make a request to the ManageIQ API

```bash
TOKEN_ENDPOINT="https://${KEYCLOAK_SERVER}:8443/auth/realms/${KEYCLOAK_REALM}/protocol/openid-connect/token"

RES=`curl -k -L --user ${USER}:${PASSWORD} -X POST \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "grant_type=password" \
   -d "client_id=${KEYCLOAK_CLIENT_ID}" \
   -d "client_secret=${OID_CLIENT_SECRET}" \
  -d "username=$USER" \
  -d "password=$PASSWORD" \
  $TOKEN_ENDPOINT`

ACCESS_TOKEN=`echo $RES | jq -r '.access_token'`

curl -L -vvv -k -X GET -H "Authorization: Bearer ${ACCESS_TOKEN}" https://${OID_CLIENT_HOST}/api/users | jq
```

3. An API Auth token

```bash
TOKEN_ENDPOINT="https://${KEYCLOAK_SERVER}:8443/auth/realms/${KEYCLOAK_REALM}/protocol/openid-connect/token"

RES=`curl -k -L --user ${USER}:${PASSWORD} -X POST \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "grant_type=password" \
  -d "client_id=${KEYCLOAK_CLIENT_ID}" \
  -d "client_secret=${OID_CLIENT_SECRET}" \
  -d "username=$USER" \
  -d "password=$PASSWORD" \
  $TOKEN_ENDPOINT`
ACCESS_TOKEN=`echo $RES | jq -r '.access_token'`

RES=`curl -L -vvv -k -X GET -H "Authorization: Bearer ${ACCESS_TOKEN}" https://${OID_CLIENT_HOST}/api/auth`
API_AUTH_TOKEN=`echo $RES | jq -r '.auth_token'`

curl -L -vvv -k -X GET  -H "Accept: application/json" -H "X-Auth-Token: ${API_AUTH_TOKEN}" https://${OID_CLIENT_HOST}/api/users | jq
```
